File: frontend/home-organizer/src/app/cookbook/models/recipe-filter.ts
import {ValInterval} from "./val-interval";
import {RecipeComp} from "./recipe-comp";
import {TSMap} from "typescript-map";
import {GroupIngredientName} from "./group-ingredient-name";

export class RecipeFilter{
  title?: string;
  kcal: ValInterval<number> = new ValInterval<number>();
  proteins: ValInterval<number> = new ValInterval<number>();
  fats: ValInterval<number> = new ValInterval<number>();
  carbons: ValInterval<number> = new ValInterval<number>();
  portions: ValInterval<number> = new ValInterval<number>();
  cookTime: ValInterval<number> = new ValInterval<number>();
  rating: ValInterval<number> = new ValInterval<number>();
  typeIds: string[] = [];
  categoryIds: string[] = [];
  kitchenIds: string[] = [];
  ingredients: Map<GroupIngredientName,RecipeComp[]> = new Map<GroupIngredientName, RecipeComp[]>();

  toJson(): any{
    let ingredientIdsMap = new TSMap<string,string[]>()
    this.ingredients.forEach((ingredients, group) =>
      ingredientIdsMap.set(group.id,ingredients.map(ingredient => ingredient.id)))
    return {
      "title": this.title,
      "kcal": this.kcal,
      "proteins": this.proteins,
      "fats": this.fats,
      "carbons": this.carbons,
      "portions": this.portions,
      "cookTime": this.cookTime,
      "rating": this.rating,
      "typeIds": this.typeIds,
      "categoryIds": this.categoryIds,
      "kitchenIds": this.kitchenIds,
      "ingredients": ingredientIdsMap
    }
  }
}

File: cookbook/src/main/java/ru/veqveq/cookbook/model/filter/RecipeFilter.java
package ru.veqveq.cookbook.model.filter;

import io.swagger.v3.oas.annotations.media.Schema;
import lombok.Data;
import ru.veqveq.cookbook.model.ValueInterval;

import java.util.List;
import java.util.UUID;

@Data
public class RecipeFilter {
    @Schema(description = "Название")
    private String title;

    @Schema(description = "Калорийность")
    private ValueInterval<Integer> kcal;

    @Schema(description = "Белки")
    private ValueInterval<Integer> proteins;

    @Schema(description = "Жиры")
    private ValueInterval<Integer> fats;

    @Schema(description = "Углеводы")
    private ValueInterval<Integer> carbons;

    @Schema(description = "Количество порций")
    private ValueInterval<Integer> portions;

    @Schema(description = "Время готовки, минут")
    private ValueInterval<Integer> cookTime;

    @Schema(description = "Рейтинг")
    private ValueInterval<Integer> rating;

    @Schema(description = "Типы блюда")
    private List<UUID> typeIds;

    @Schema(description = "Категории блюда")
    private List<UUID> categoryIds;

    @Schema(description = "Кухни")
    private List<UUID> kitchenIds;

    @Schema(description = "Названия ингредиентов")
    private List<UUID> ingredientNameIds;

    @Schema(description = "Названия обобщенных ингредиентов")
    private List<UUID> groupIngredientNameIds;
}
